dropshot/src/api_description.rs

```diff
@@ -972,34 +979,74 @@ impl<Context: ServerContext> ApiDescription<Context> {
     }
 }
 
+/// A collection of errors that occurred while building an `ApiDescription`.
+///
+/// Returned by the `api_description` and `stub_api_description` functions
+/// generated by the [`api_description`](crate::api_description) macro.
 #[derive(Debug, Clone, PartialEq, Eq)]
-pub struct ApiDescriptionBuildError {
-    errors: Vec<String>,
+pub struct ApiDescriptionBuildErrors {
+    errors: Vec<ApiDescriptionRegisterError>,
 }
 
-impl ApiDescriptionBuildError {
-    /// Create a new `ApiDescriptionBuildError` with the given errors.
-    pub fn new(errors: Vec<String>) -> Self {
+impl ApiDescriptionBuildErrors {
+    /// Create a new `ApiDescriptionBuildErrors` with the given errors.
+    pub fn new(errors: Vec<ApiDescriptionRegisterError>) -> Self {
         Self { errors }
     }
 
-    /// Return a list of the errors that occurred while building an
-    /// `ApiDescription`.
-    pub fn errors(&self) -> &[String] {
+    /// Return a list of the errors that occurred.
+    pub fn errors(&self) -> &[ApiDescriptionRegisterError] {
         &self.errors
     }
 }
 
-impl fmt::Display for ApiDescriptionBuildError {
+impl fmt::Display for ApiDescriptionBuildErrors {
     fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
         write!(f, "failed to register endpoints: \n")?;
         for error in &self.errors {
-            write!(f, "  - {}\n", error)?;
+            write!(
+                f,
+                "  - registering '{}' failed: {}\n",
+                error.operation_id, error.message
+            )?;
         }
         Ok(())
     }
 }
 
+impl std::error::Error for ApiDescriptionBuildErrors {}
+
+/// An error that occurred while registering an individual API endpoint.
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct ApiDescriptionRegisterError {
+    operation_id: String,
+    message: String,
```

Comment:
Rather than storing an unstructured string, we could model the different kinds of errors as an enum if that makes sense. That would also let tests not have to do the somewhat brittle string matching they have to do today.

Reply:
We had a meeting yesterday and we decided to not do this for now -- tracking endpoint names (operation IDs) in a structured manner is very useful, but the actual message less so. But we can definitely do this in the future if required.

Reply:
I think both statements are true here: enums are generally better **and** in this case, no one is going to unpick these messages programmatically. It is definitely an improvement to have the operations as explicit strings.

Reply:
Agreed on both counts.